test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp

```diff
@@ -0,0 +1,364 @@
+ * constraints, c1 and c2). For generated constraints that are not compatible
+ * with the supplied constraints, the generate_samples callable is expected
+ * to throw domain_errors.
+ *
+ * The generate_samples callable is also passed various other guaranteed invalid
+ * values like positive infinity, negative infinity, and NaNs (these should also
+ * cause domain_errors).
+ *
+ * The generated_samples callable is also tested to reject incompatibly sized
+ * input vectors. These should cause invalid_argument errors.
+ *
+ * The generate_samples callable must have the signature:
+ *   T_out generate_samples(T_param1 p1, T_param2 p2, T_rng)
+ * where T_param1 and T_param2 can be any combination of:
+ *   double, std::vector<double>, VectorXd, or RowVectorXd
+ * and T_rng is a boost random number generator
+ *
+ * The output of generate_samples must be of
+ *   stan::length(out) == stan::max_size(p1, p2)
+ *
+ * Each element of the output, out[i], should correspond to a random variate
+ * generated according to parameters p1[i] and p2[i]
+ *
+ * @tparam T_param1 Type of first parameter
+ * @tparam T_param2 Type of second parameter
+ * @tparam T_generate_samples Type of generate_samples functor
+ * @param generate_samples generate_samples functor to test
+ * @param c1 Constraint of first parameter
+ * @param c2 Constraint of second parameter
+ */
+template<typename T_param1, typename T_param2, typename T_generate_samples>
+void twoParamDistCheckThrows(T_generate_samples generate_samples, Constraint c1,
+                             Constraint c2) {
+  boost::random::mt19937 rng;
+
+  T_param1 p1;
+  T_param2 p2;
+
+  std::vector<Constraint> constraints = { Constraint::Positive,
+                                          Constraint::NonNegative,
+                                          Constraint::None };
+
+  for (auto it1 = constraints.begin(); it1 != constraints.end(); it1++) {
+    for (auto it2 = constraints.begin(); it2 != constraints.end(); it2++) {
+      generateValues(p1, *it1, 3);
+      generateValues(p2, *it2, 3);
+
+      if (compatible(*it1, c1) && compatible(*it2, c2)) {
+        EXPECT_NO_THROW(generate_samples(p1, p2, rng));
+      } else {
+        EXPECT_THROW(generate_samples(p1, p2, rng), std::domain_error);
+      }
+    }
+  }
+
+  std::vector<double> badValues = { stan::math::positive_infinity(),
+                                    stan::math::negative_infinity(),
+                                    stan::math::not_a_number() };
+
+  for (auto it = badValues.begin(); it != badValues.end(); it++) {
+    generateValues(p1, c1, 3);
+    generateValues(p2, c2, 3);
+    setValue(p1, *it);
+    EXPECT_THROW(generate_samples(p1, p2, rng), std::domain_error);
+
+    generateValues(p1, c1, 3);
+    setValue(p2, *it);
+    EXPECT_THROW(generate_samples(p1, p2, rng), std::domain_error);
+  }
+
+  generateValues(p1, c1, 3);
+  generateValues(p2, c2, 4);
+  if (stan::length(p1) != 1 && stan::length(p2) != 1)
```

Comment:
It's my high quality return-value-as-a-parameter code :(.

```
generateValues(p1, c1, 3); // Make p1 length 3 with constraint c1 if it's vector-like otherwise just assign a single value
generateValues(p2, c2, 4); // Make p2 length 4 with constraint c2 if it's vector-like otherwise just assign a single value
```